### ISSUE TITLE
Segmentation fault when encountering a TemplateHaskell splice

### ISSUE THREAD

**Issue by istathar** [haskell/haskell-language-server#800]
We've had **ghcide** crashing, and the editor trying to reload it leads to a tight loop that the editor can't escape. Trying the _ghcide_ binary from the command-line in test mode leads to a segfault:

```
Step 6/6: Type checking the files
Segmentation fault (core dumped)
```

This is triggered when using TemplateHaskell. We have a splice [`fromPackage :: Q Exp`](https://hackage.haskell.org/package/core-program-0.2.4.2/docs/Core-Program-Metadata.html#v:fromPackage) which returns some metadata from the project's _.cabal_ file:

```haskell
version :: Version
version = $(fromPackage)
```

removing this definition from the source file (in this case [_RenderMain.hs_](https://github.com/oprdyn/publish/blob/master/src/RenderMain.hs#L13-L14) lines 13-14) allows the _ghcide_ binary to complete its test run:

```
Step 6/6: Type checking the files

Completed (1 file worked, 0 files failed)
```
I had earlier speculated that QuasiQuotes were to blame, but the file checks ok with those `[quote| ... |]` blocks still in the file.

Other files in the project load and check fine, so long as you avoid having the problematic file open in an editor tab you can avoid the crash. Not ideal, but at least a workaround.

**Comment by FinleyMcIlwaine** [comment]
I'm also running into this with a cabal project on MacOS. The specific thing that's causing it for me is a typed TH quotation. Using GHC 8.10.2 and HLS 0.9.0. I can provide more information if it may help lead to a fix!

**Comment by mpickering** [comment]
@FinleyMcIlwaine That is probably unrelated to this problem. Please make a new ticket.

**Comment by andreasewering** [comment]
I'm running into a segfault problem as well using postgresql-typed's template haskell. Tried with GHC 8.8.4, 8.10.2 and 8.10.4.
https://github.com/haskell/haskell-language-server/issues/1297#issuecomment-821837822
Intellij Haskell does not seem to have problems with this, probably because it isn't ghcide-based?

**Comment by jneira** [comment]
> Intellij Haskell does not seem to have problems with this, probably because it isn't ghcide-based?

Yeah it does not use ghcide, comparing with intellij could be a investigation path, thanks

**Comment by NSilv** [comment]
Just chiming in to say I get segfaults as well, only with template haskell on (specifically, quasiquotes from the path library). Although in my case, it only happens inside a nix-shell. (GHC 8.10.4, on WSL2+Fedora)

**Comment by franleplant** [comment]
I got this same problem and the only way I got to make it work was running `hls` inside `nix-shell`, in particular, I am developing plutus smart contracts and so I used the `nix-shell` config for `plutus`, find it [here](https://github.com/input-output-hk/plutus/blob/master/default.nix)

**Comment by ZabelTech** [comment]
Hi, I got the same problem with HLS 1.3.0.0 and GHC 8.8.4.
HLS 1.1.0.0 and GHC 8.10.4 works fine.

**Comment by vst** [comment]
+1

```
HLS: 1.4.0.0
GHC: 8.10.7
Stackage LTS: lts-18.10
```

This one is related to [file-embed](https://hackage.haskell.org/package/file-embed) library usage. Plain vanilla usage works:

```hs
$(makeRelativeToProject` "<relative-path-to-file>" >>= embedStringFile)
```

... but this helper causes issues (I guess at the call-site):

```hs
includeFile :: FilePath -> Q Exp
includeFile = makeRelativeToProject >=> embedStringFile
```

This was a quick hack to mitigate #481. So I am not sure if it is separate from #481, but thought I should report.

Also, in the past, I experienced issues with quasiquotes from the `path` library as reported earlier. I did not check with the recent versions, though.

**Comment by DylanSp** [comment]
I'm running what I _suspect_ might be this problem when trying to use HLS and VS Code on [Wasp's Haskell component](https://github.com/wasp-lang/wasp/tree/master/waspc). The symptoms I see in VS Code are a crash loop, similar to https://github.com/haskell/haskell-language-server/issues/1297, with no errors reported from the server in the VS Code output.

When I run HLS 1.4.0 from an isolated install (installed with `ghcup install hls --isolate ~/ghcup-isolated/hls`, ran from the waspc directory with `~/ghcup-isolated/hls/haskell-language-server-8.10.4 .`), I get a segfault, which is why I think this issue might be the problem. I tried inspecting the core dump with gdb, but backtracing didn't show any source info. If there's any other analysis I can run on the core dump, let me know, it's not my area of expertise.

I tried to replicate by building from source; I cloned this repo, and checked out the 1.4.0 tag. Building with Stack with or without debug symbols, with or without optimization, I wasn't able to replicate the issue; running the built executable against waspc ran to completion every time. I tried:
- `stack build`
- `stack build --ghc-options="-O2"`
- `stack build --no-strip --ghc-options="-g"`
- `stack build --no-strip --ghc-options="-g O2"`

I haven't tried building from source with Cabal yet. I don't know if there are any other options/flags I should use to replicate what's used for the binary fetched by `ghcup`.

**Comment by jneira** [comment]
@DylanSp many thanks for the detailed update and for trying reproduce the problem.
It seems build from source fix the issue so i guess linking locally the needed libraries are the cause of it: what is your os?
It would be great to confirm building from source is a workaround for this.
Nowadays ghcup allows you building hls from source (see `ghcup compile hls --help`) just in case you could find it useful

**Comment by DylanSp** [comment]
@jneira My OS is Linux Mint 18, based on Ubuntu 16.04, running on x64. It's entirely possible that something's weird with linking with local libraries; I had errors building an unrelated Haskell project due to GHC 8.10.7/base 4.10.3.0 wanting a newer `glibc` than what my OS provides. I'll take a stab at using ghcup to compile from source and see if I can replicate the issue that way.

**Comment by DylanSp** [comment]
Tried building HLS from source with `ghcup compile`. That required installing GHC 8.10.4 first with `ghcup install ghc 8.10.4`; that failed due to 
```
/lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.27' not found (required by libraries/base/dist-install/build/libHSbase-4.14.1.0-ghc8.10.4.so)
```
Will try using a Docker container to build against a more recent libc.

**Comment by jneira** [comment]
The prebuilt binary is being built with ubuntu 18 so it seems is could be related, other issues about crashes with TH are not reproduced building hls from source so i think they are related:
* #1982 (ubuntu 20)
* #469 (macos)

**Comment by DylanSp** [comment]
I was able to reproduce the core dump when installing the prebuilt HLS binary in an Ubuntu 18.04 container. I tried building HLS from source in an 18.04 container as well; that ran successfully to completion. This C assertion failure fired after HLS finished analyzing everything:
```
[...]
Completed (179 files worked, 3 files failed)
haskell-language-server-8.10.4: allocatestack.c:384: advise_stack_range: Assertion `freesize < size' failed.

Aborted (core dumped)
```

https://gist.github.com/DylanSp/aea8643a19e883a19f974e652aada853 can reproduce the error.
```
$ docker build -t hls_prebuilt:18.04 .
$ docker run -it hls_prebuilt:18.04 bash
root@[container ID]:/wasp/wasp/waspc# $ stack build
root@[container ID]:/wasp/wasp/waspc# $ /hls_isolated/haskell-language-server 8.10.4 .
```

**Comment by DylanSp** [comment]
After further investigation, I was able to get the same assertion failure with HLS built from source, in both Ubuntu 18.04 and 20.04 environments. The problem seems to be related to the prebuilt HLS binary using static linking; the problem didn't occur when HLS was built without the `--enable-executable-static` flag. Furthermore, when building with that flag enabled, the linker reported several warnings about needing " at runtime the shared libraries from the glibc version used for linking" for various functions. Dockerfiles for reproduction and the full text of the linker warnings are in https://gist.github.com/DylanSp/19a3b076bbf53320cbae5573790cd50e.

**Comment by jneira** [comment]
@DylanSp many thanks for the anaylisis and reproduction, so we could state that building from source with dynamic linking seems to workaround the issue (as other ones): if any one of the reporters can reproduce it with that setup, please update the issue, thanks!

**Comment by DylanSp** [comment]
@jneira No problem; glad I can help by suggesting a workaround, and I hope with my info you can track down the root cause.

**Comment by jneira** [comment]
This issue was really useful to make progress on diagnostic template haskell related problems, thanks very much all for the effort put in investigations and reproductions.
I think we can close it as we have newer issues with more up to date info about the possible fix and the dynamic workaround (see #1431 and #2000)

**Comment by jneira** [comment]
I am gonna close this issue as all compiler crashes seems to have the same root cause:
* We have a general issue about: https://github.com/haskell/haskell-language-server/issues/1431
* And other one with more info about the causes: https://github.com/haskell/haskell-language-server/issues/2000
* We have a generic workaround (build hls from source dynamically linked) documented in the troubleshooting guide: https://haskell-language-server.readthedocs.io/en/latest/troubleshooting.html#static-binaries-and-template-haskell-support
* And a possible fix upstream for ghc described in https://github.com/haskell/haskell-language-server/issues/2000#issuecomment-974658606

If any of you think the issue should not be included generically feel free to reopen it (with a brief explanation if possible)
Thanks all!
